core/app/models/spree/order_contents.rb

```diff
@@ -83,6 +84,20 @@ def reload_totals
       @order.recalculate
     end
 
+    def destroy_non_positive_line_items(params)
+      line_items_to_destroy_ids = []
+      # Could be a single LineItem or multiple via nesting
+      if id = params[:line_items_attributes][:id].present?
+        line_items_to_destroy_ids << id if params[:line_items_attributes][:quantity].to_i <= 0
+      else
+        params[:line_items_attributes].reject! do |_index, attributes|
+          line_items_to_destroy_ids << attributes[:id] if attributes[:quantity].to_i <= 0
+        end
+      end
+      order.line_items.where(id: line_items_to_destroy_ids).destroy_all
```

Comment:
Thoughts on using https://github.com/solidusio/solidus/blob/df68c0c748ded6a595d28c37a9460126c6d12ecb/core/app/models/spree/order_contents.rb#L32?

Reply:
I would prefer not to use this method because it would cause a DELETE SQL for every destroy rather than just one. It would also cause two extra order recalculates for every destroy. Note that `update_cart` explicitly handles recalculating anyway.